### PR TITLE
ApplicationBuilder

### PR DESCRIPTION
**Pull request by Valenzione**
I've realized that we have no way of creating **new** Applications from `hydrosdk`. I've drafted how we can introduce `ExecutionGraph`, `ExecutionStage`, and most importantly `ApplicationBuilder`. 

This is only a quick draft, but IMO it'll increase code quality overall (due to replacing passing dicts around with well-defined objects) as well as adding new functionality. I think we could spend a few days on that closer to the 2.4 release date. WDYT guys? @trthhrtz @KineticCookie @mkf-simpson

### CODE REVIEW

**Comment by trthhrtz** [comment]
'good first issue' label?

**Comment by Valenzione** [comment]
Following classes were created:
* Application
* ApplicationBuilder - to ease the creation of applicaction 
* ExecuionGraph - list of execution stages
* ExecutionStage -  list of ModelVariants
* ExecutionStageBuilder - to ease the creation of ExecutionStages
* ModelVariant = `namedtuple(ModelVersion, weight)`

example of using new improvements:

```python
cluster = Cluster('HTTP_CLUSTER_ADDRESS')

m1 = ModelVersion.find(cluster, "census_income_classification", 9)
m2 = ModelVersion.find(cluster, "census_income_classification", 10)

application = ApplicationBuilder("census_ab_test").with_stage(
    ExecutionStageBuilder().with_model_variant(m1, 50).
        with_model_variant(m2, 50).build()).build(cluster)
```

**Comment by Valenzione** [comment]
@trthhrtz @KineticCookie I'll merge updates from the master after your review since there are a lot of updates from the `/docs` folder which may distract you from this review

File: hydrosdk/application.py [hunk 1]
```diff
@@ -33,250 +21,262 @@ class ApplicationStatus(Enum):
...
-        resp = cluster.request(method="POST", url="/api/v2/application", json=application)
         if resp.ok:
             resp_json = resp.json()
-            app = Application.app_json_to_app_obj(cluster=cluster, application_json=resp_json)
+            app = Application.from_json(cluster=self.cluster, application_json=resp_json)
             return app
 
-        raise Exception(
-            f"Failed to create application. Application = {application}. {resp.status_code} {resp.text}")
+        raise Exception(f"Failed to create application {self.name}. {resp.status_code} {resp.text}")
```
Comment (KineticCookie) [inline hydrosdk/application.py:203]:
```suggestion
        raise ApplicationCreationError(f"Failed to create application {self.name}. {resp.status_code} {resp.text}")
```

File: hydrosdk/application.py [hunk 2]
```diff
@@ -33,250 +21,262 @@ class ApplicationStatus(Enum):
...
+
+    def delete(self, ):
+        resp = self.cluster.request("DELETE", f"{Application._BASE_URL}/{self.name}")
         if resp.ok:
-            resp_json = resp.json()
-            applications = [Application.app_json_to_app_obj(cluster=cluster, application_json=app_json) for app_json in
-                            resp_json]
-            return applications
+            return True
+        raise Exception(f"Failed to delete application. Name = {self.name}. {resp.status_code} {resp.text}")
```
Comment (KineticCookie) [inline hydrosdk/application.py:135]:
```suggestion
        raise ApplicationDeletionError(f"Failed to delete application. Name = {self.name}. {resp.status_code} {resp.text}")
```

File: hydrosdk/application.py [hunk 3]
```diff
@@ -33,250 +21,262 @@ class ApplicationStatus(Enum):
...
+    def list_all(cluster) -> List['Application']:
+        resp = cluster.request("GET", Application._BASE_URL)
+        if resp.ok:
+            applications = [Application.from_json(cluster=cluster, application_json=app_json) for app_json in resp.json()]
+            return applications
 
-    def predictor(self, return_type=PredictorDT.DICT_NP_ARRAY) -> PredictServiceClient:
-        self.impl = MonitorableImplementation(channel=self.cluster.channel, target=self.name)
-        self.predictor_return_type = return_type
+        raise Exception(
```
Comment (KineticCookie) [inline hydrosdk/application.py:33]:
```suggestion
        raise BadResponse(
```

or something like that

File: hydrosdk/application.py [hunk 4]
```diff
@@ -33,250 +21,262 @@ class ApplicationStatus(Enum):
...
-        self.status = application.status
+        raise Exception(f"Failed to find by name. Name = {name}. {resp.status_code} {resp.text}")
 
     @staticmethod
-    def app_json_to_app_obj(cluster: Cluster, application_json: dict) -> 'Application':
+    def delete(cluster, name):
+        resp = cluster.request("DELETE", f"{Application._BASE_URL}/{name}")
+        if resp.ok:
+            return True
+        raise Exception(f"Failed to delete application. Name = {name}. {resp.status_code} {resp.text}")
```
Comment (KineticCookie) [inline hydrosdk/application.py:51]:
```suggestion
        raise ApplicationDeletionError(f"Failed to delete application. Name = {name}. {resp.status_code} {resp.text}")
```

File: hydrosdk/application.py [hunk 5]
```diff
@@ -33,250 +21,262 @@ class ApplicationStatus(Enum):
...
+            return app
 
-    def update_status(self) -> None:
-        """
-        Setter method that updates application status
-        :return: None
-        """
-        application = Application.find_by_name(cluster=self.cluster, app_name=self.name)
-        self.status = application.status
+        raise Exception(f"Failed to find by name. Name = {name}. {resp.status_code} {resp.text}")
```
Comment (KineticCookie) [inline hydrosdk/application.py:44]:
```suggestion
        raise ApplicationNotFoundError(f"Failed to find by name. Name = {name}. {resp.status_code} {resp.text}")
```

**Comment by tidylobster** [comment]
👍

**Pull request merged**